diffkemp/simpll/Logger.h

```diff
@@ -0,0 +1,149 @@
+// found in both this comparison and all its predecessors.
+#define RETURN_WITH_LOG(return_value)                                          \
+    do {                                                                       \
+        const auto &x = return_value;                                          \
+        LOG_KEEP(x);                                                           \
+        return x;                                                              \
+    } while (false)
+
+// Force variant of RETURN_WITH_LOG.
+// If LOGGER_FORCE_LEVEL is not enabled, behaves identically to RETURN_WITH_LOG.
+// If it is enabled, forces keeping the message regardless of the given value
+// and marks that message as force-kept. However, its children that weren't
+// previously marked as force-kept will still be erased if the given value is
+// evaluated to false.
+// Erasing non-force-kept children prevents logging of unforced comparisons
+// where a difference was found, but no difference was found in at least one of
+// its (forced) predecessor comparisons.
+// Used for comparisons where it is be desirable to log that a comparison has
+// occured even if no difference was found (when using higher level of debug
+// verbosity).
+#define RETURN_WITH_LOG_FORCE(return_value)                                    \
+    do {                                                                       \
+        const auto &x = return_value;                                          \
+        LOG_KEEP_FORCE(x);                                                     \
+        return x;                                                              \
+    } while (false)
+
+#define LOG_NO_INDENT(msg) DEBUG_WITH_TYPE(DEBUG_SIMPLL, dbgs() << msg)
+#define LOG_VERBOSE_NO_INDENT(msg)                                             \
+    DEBUG_WITH_TYPE(DEBUG_SIMPLL_VERBOSE, dbgs() << msg)
+#define LOG_VERBOSE_EXTRA_NO_INDENT(msg)                                       \
+    DEBUG_WITH_TYPE(DEBUG_SIMPLL_VERBOSE_EXTRA, dbgs() << msg)
+
+#define LOG(msg) LOG_NO_INDENT(getDebugIndent() << msg)
+#define LOG_VERBOSE(msg) LOG_VERBOSE_NO_INDENT(getDebugIndent() << msg)
+#define LOG_VERBOSE_EXTRA(msg)                                                 \
+    LOG_VERBOSE_EXTRA_NO_INDENT(getDebugIndent() << msg)
+
+#define LOG_INDENT() DEBUG_WITH_TYPE(DEBUG_SIMPLL, increaseDebugIndentLevel())
+#define LOG_UNINDENT() DEBUG_WITH_TYPE(DEBUG_SIMPLL, decreaseDebugIndentLevel())
+
+#define WITHOUT_LOGGER_INDENT(actions)                                         \
+    do {                                                                       \
+        size_t indent;                                                         \
+        DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, indent = logger.resetIndent());     \
+        actions;                                                               \
+        DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, logger.loadIndent(indent));         \
+    } while (false)
```

Comment:
What exactly is the purpose of this macro? Does it have some future use? (currently it's not used anywhere)